src/util/common_table.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * common_table.c
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <pelz_io.h>
+#include <common_table.h>
+#include <util.h>
+#include <pelz_request_handler.h>
+#include <charbuf.h>
+#include <pelz_log.h>
+
+#include "sgx_trts.h"
+#include "pelz_enclave_t.h"
+#include "kmyth_enclave_trusted.h"
+
+Table key_table = {
+  .entries = NULL,
+  .num_entries = 0,
+  .mem_size = 0
+};
+
+Table server_table = {
+  .entries = NULL,
+  .num_entries = 0,
+  .mem_size = 0
+};
+
+//Destroy server table
+int table_destroy(int type)
+{
+  Table table;
+
+  pelz_log(LOG_DEBUG, "Table Destroy Function Starting");
+
+  switch (type)
+  {
+  case KEY:
+    table = key_table;
+    break;
+  case SERVER:
+    table = server_table;
+    break;
+  default:
+    return (1);
+  }
+
+  for (unsigned int i = 0; i < table.num_entries; i++)
+  {
+    if (table.entries[i].id.len != 0)
+    {
+      free_charbuf(&table.entries[i].id);
+    }
+    if (type == KEY)
+    {
+      if (table.entries[i].value.cert.len != 0)
+      {
+        secure_free_charbuf(&table.entries[i].value.key);
+      }
+    }
+    if (type == SERVER)
+    {
+      if (table.entries[i].value.cert.len != 0)
+      {
+        secure_free_charbuf(&table.entries[i].value.cert);
+      }
+    }
+  }
+
+  //Free the storage allocated for the hash table
+  free(table.entries);
+  table.entries = NULL;
+  table.num_entries = 0;
+  table.mem_size = 0;
+
+  if (type == KEY)
+  {
+    key_table = table;
```

Comment:
I'm not sure I understand what this assignment, or the assignment on line 85 are doing. ```table``` is already the correct one of ```key_table``` or ```server_table``` from the switch on line 39. I don't see anything in the code that would cause ```table``` to not be the correct structure here already.

Reply:
""table"" is a local variable and ''"key_table"" or ""server_table"" are global variables so the changes to ""table"" will not stay after the function if you don't reassign the changes back to the global variable.

Reply:
In that case I suggest making ```table``` a pointer to one of ```key_table``` or ```server_table``` so you can avoid the re-assignment.